libraries/c_sdk/standard/https/src/private/iot_https_internal.h

```diff
@@ -0,0 +1,506 @@
+ * in the header buffer.
+ * This state is used in the parser callback _httpParserOnHeaderFieldCallback() to check if the current header field
+ * parsed equals the header we are searching for. It is used in parser callback _httpParserOnHeaderValueCallback() to
+ * return the header value if the corresponding field we are searching for was found. It is used in parser callback
+ * _httpParserOnHeadersCompleteCallback() to stop parsing the header buffer if the header we are searching for was not
+ * found.
+ *
+ * The header buffer is separate from the body buffer.
+ * The header buffer is configured in #IotHttpRequestInfo_t.respUserBuff. The body buffer is configured in
+ * #IotHttpRequestInfo_t.syncInfo->respData or as buffer provided asynchronously during the
+ * #IotHttpsClientCallbacks_t.readReadyCallback() to call to @ref IotHttpsClient_ReadResponseBody().
+ */
+typedef enum IotHttpsResponseBufferState
+{
+    PROCESSING_STATE_NONE,                   /**< @brief There is no buffer processing currently. */
+    PROCESSING_STATE_FILLING_HEADER_BUFFER,  /**< @brief The header buffer is being filled and parsed. */
+    PROCESSING_STATE_FILLING_BODY_BUFFER,    /**< @brief The body buffer is being filled and parsed. */
+    PROCESSING_STATE_FINISHED,               /**< @brief Filling and parsing of both buffers is finished. */
+    PROCESSING_STATE_SEARCHING_HEADER_BUFFER /**< @brief The header buffer is being searched. */
+} IotHttpsResponseBufferState_t;
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Represents an HTTP connection.
+ */
+typedef struct _httpsConnection
+{
+    const IotNetworkInterface_t * pNetworkInterface; /**< @brief Network interface with calls for connect, disconnect, send, and receive. */
+    void * pNetworkConnection;                       /**< @brief Pointer to the network connection to use pNetworkInterface calls on. */
+    uint32_t timeout;                                /**< @brief Timeout for a connection and waiting for a response from the network. */
+
+    /**
+     * @brief true if a connection was successful most recently on this context
+     *
+     * We have no way of knowing if the server closed the connection because that error is unique to the underlying TLS
+     * layer. This is set to false initially, then set to true for a successful intentional call to connect.
+     * Post connection, this is set to false only after an implicit disconnect with a non-persistent request, an implicit
+     * disconnect with a network error, or an explicit disconnect with a call to @ref https_client_function_disconnect.
+     */
+    bool isConnected;
+    bool isDestroyed; /**< @brief true if the connection is already destroyed and we should call anymore  */
+
+    /**
+     * @brief true if the a connection is in the process of disconnecting.
+     *
+     * Mutexes are large and expensive, so it is better to just check a variable then return.
```

Comment:
This comment is confusing because the it implies that thread safety is required, whereas this approach is not thread safe.

Reply:
Just realized that a thread can context switch in the middle of:
```
    /* If this routine is currently is progress by another thread, for instance the taskpool worker that received a
     * network error after sending, then return right away because connection resources are being used. */
    if( connHandle->isDisconnecting )
    {
        HTTPS_SET_AND_GOTO_CLEANUP( IOT_HTTPS_BUSY );
    }
    else
    {
        connHandle->isDisconnecting = true;
    }
```
And see isDisconnecting as false twice.
So I will switch to a disconnection mutex.